flink-table/flink-table-planner/src/main/java/org/apache/flink/table/planner/plan/nodes/exec/ExecNodeBase.java

```diff
@@ -49,6 +49,13 @@
 @JsonIgnoreProperties(ignoreUnknown = true)
 public abstract class ExecNodeBase<T> implements ExecNode<T> {
 
+    /**
+     * The default value is chosen for the JSON deserialization case. Other cases must set this flag
+     * accordingly via {@link #setCompiled(boolean)}. It is not exposed via a constructor arg to
+     * avoid complex constructor overloading for all {@link ExecNode}s.
+     */
+    private transient boolean isCompiled = true;
```

Comment:
Would it be useful to not have a general default and force all manual constructions to set this, to make sure this isn't accidentally missed?

For the Json deserialization case you could try adding an `StdConverter` with `@JsonDeserialize` that sets this to `true` post deserialization.

```
@JsonDeserialize(converter = ExecNodeBase.IsCompiledFlagDefaultSetter.class)
public abstract class ExecNodeBase<T> implements ExecNode<T> {

    public static class IsCompiledFlagDefaultSetter extends StdConverter<ExecNodeBase<?>, ExecNodeBase<?>> {
        @Override
        public ExecNodeBase<?> convert(ExecNodeBase<?> execNodeBase) {
            execNodeBase.setCompiled(true);
            return execNodeBase;
        }
    }
```

Reply:
The more time you invest in this topic, the nicer the solution gets. I found a better way using `@JsonInject`. I will merge this once the build turns green, if you don't have any objections.